### PR TITLE
setup: migrate to jupyter server 2

### LINKED ISSUE #411: jupyter_server: support version 2
This is currently blocked by outdated and cylc-flow dependency on `pyzmq == 22.*`.

Upgrading the cylc-flow dependency will allow jupyter_server 2 to be used.

Jupyter server 2 does not support 3.7 so we either need to test for both or drop 3.7 support.

**Pull requests welcome!**

### PR DESCRIPTION
**Pull request by oliver-sanders**
Closes #411

Migrate from Jupyter Server 1-2.

There has been a rejig of the `get_current_user` logic so this should be accompanied by a security check of authorisation/authentication.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/625.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.
- [x] Security penetration test.

### CODE REVIEW

**Comment by hjoliver** [comment]
Yes, it's ready for review now, and not a big change. It should be merged pretty soon.

**Review by hjoliver — COMMENTED** [review]
Looks good at first glance. I haven't tested it yet.

**Comment by oliver-sanders** [comment]
@mahendrapaipuri, we should be able to get this reviewed soonish, however, this feature is scheduled for release along with Cylc 8.2.0 so won't be released until some other work is completed.

If you want to get going with Jupyter Server 2 and Cylc 8.1, we might be able to provide you with a development branch you could install from to save waiting.

**Comment by mahendrapaipuri** [comment]
Cheers @oliver-sanders!! Well, we are not in a rush. I will keep track of this PR and once it gets merged I can test using the `master` branch on our platform.

**Comment by oliver-sanders** [comment]
Great, just be aware you'll need to install cylc-flow from master too.

File: setup.cfg
Comment (MetRonnie) [inline setup.cfg:1]: Also need to update 
https://github.com/cylc/cylc-uiserver/blob/0f4500548b6181df16dbf2779f79cce15b639888/setup.cfg#L36
https://github.com/cylc/cylc-uiserver/blob/0f4500548b6181df16dbf2779f79cce15b639888/setup.cfg#L44

**Review by MetRonnie — CHANGES_REQUESTED** [review]

File: setup.cfg
Comment (mahendrapaipuri) [reply setup.cfg:1]: Probably a relevant comment here. Seems like [cylc-flow](https://github.com/cylc/cylc-flow/blob/f81ebc298f8f9514ad588c12da480815d4b286cd/setup.cfg#L53-L54) supports till Python 3.11 and [cylc-uiserver](https://github.com/cylc/cylc-uiserver/blob/ed7f6899132392f6edb9f9f072bc54db92dcd9a8/setup.cfg#L38) only until Python 3.9. Are these classifiers up to date? I mean it is a quite trivial use case that we use them together, right?

Comment (MetRonnie) [reply setup.cfg:1]: See #388 for why uiserver support is limited to < 3.10

**Review by MetRonnie — CHANGES_REQUESTED** [review]
Still need to update to `>=3.8`

Or should it be `>=3.8,<3.10`?
https://github.com/cylc/cylc-uiserver/blob/14467609cc3480d0487e3ea8d4294ce03accbdbd/setup.cfg#L43

**Comment by oliver-sanders** [comment]
sorry, didn't read properly, done.

**Review by MetRonnie — APPROVED** [review]
LGTM, I've done a basic test of running the UI after pip installing on this branch

**Comment by wxtim** [comment]
- [x] Look at code
- [x] Have a good play with Cylc Hub
- [ ] Pen test

File: setup.cfg [hunk 1]
```diff
@@ -54,8 +53,7 @@ install_requires =
     ansimarkup>=1.0.0
     graphene
     graphene-tornado==2.6.*
-    graphql-ws==0.4.4
-    jupyter_server>=1.10.2, <2.0
+    jupyter_server>=2.0
```
Comment (oliver-sanders) [inline setup.cfg:56]:
```suggestion
    graphql-ws==0.4.4
    jupyter_server>=2.0
```

File: setup.cfg [hunk 2]
```diff
@@ -54,8 +53,7 @@ install_requires =
     ansimarkup>=1.0.0
     graphene
     graphene-tornado==2.6.*
-    graphql-ws==0.4.4
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply setup.cfg:57]: Cheers.

**Review by MetRonnie — APPROVED** [review]

**Comment by oliver-sanders** [comment]
Ok, this is now looking ok.

This PR now upgrades us, both to Jupyter Server v2 and to newer versions of Jupyter Hub which has required some re-working of the authorisation which will need to be carefully examined and tested.

The good news is that recent Jupyter Hub versions have added a scope to authorise the starting of other users servers so this improves multi-user functionality greatly.

Along with the [documentation PR](https://github.com/cylc/cylc-doc/pull/625):

* Partially addresses #298
* Closes #275

TODO:

* [ ] Fix tests.
* [ ] Check jupyter_server version range (we may need to up the minimum).
* [ ] Decide the jupyter_hub version range.
* [ ] Further penetration testing of Cylc UIServer and Jupyter Lab according to the [documented configurations](https://github.com/cylc/cylc-doc/pull/625).
* [ ] Add details in the [changes](https://cylc.github.io/cylc-doc/stable/html/reference/changes.html) section of the docs

**Review by MetRonnie — COMMENTED** [review]
Done another review. No comments at the moment, pending changes

**Review by hjoliver — APPROVED** [review]
LGTM, tested as working.

**Pull request merged**